### PR TITLE
Fix specified CPUs in piped commands

### LINKED ISSUE #750: Fix number of CPUs in piped commands
Basically this: https://github.com/nf-core/modules/pull/499

### PR DESCRIPTION
**Pull request by jfy133**
To close #750 

## PR checklist

- [x] This comment contains a description of changes (with reason).
- [x] If you've fixed a bug or added code that should be tested, add tests!
  - [ ] If you've added a new tool - add to the software_versions process and a regex to `scrape_software_versions.py`
  - [ ] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](https://github.com/nf-core/eager/tree/master/.github/CONTRIBUTING.md)
  - [ ] If necessary, also make a PR on the nf-core/eager _branch_ on the [nf-core/test-datasets](https://github.com/nf-core/test-datasets) repository.
- [ ] Make sure your code lints (`nf-core lint .`).
- [ ] Ensure the test suite passes (`nextflow run . -profile test,docker`).
- [x] Usage Documentation in `docs/usage.md` is updated.
- [x] Output Documentation in `docs/output.md` is updated.
- [x] `CHANGELOG.md` is updated.
- [x] `README.md` is updated (including new tool citations and authors/contributors).

### CODE REVIEW

**Comment by mahesh-panchal** [comment]
Is there a check in the code somewhere that `cpus` is at minimum 2 and at least divisible by 2 (since it's used to calculate `split_cpus`)?

**Comment by jfy133** [comment]
> Is there a check in the code somewhere that `cpus` is at minimum 2 and at least divisible by 2 (since it's used to calculate `split_cpus`)?

There isn't a specific check for (1): the number of CPUs is specified with the `label` parameter of the process, so assuming users don't modify that (and I don't really see why they would), the split_cpu varliable is only assigned to multi-threaded label'd processes.

And for (2) I believe the `floor` part of the split_cpu definition is account for the divisible by 2 issue

File: main.nf [hunk 1]
```diff
@@ -1593,26 +1596,26 @@ process samtools_filter {
         '''
     } else if ( "${params.bam_unmapped_type}" == "bam" && params.bam_filter_minreadlength != 0 ){
         '''
-        samtools view -h !{bam} | samtools view - -@ !{task.cpus} -f4 -o !{libraryid}.unmapped.bam
-        samtools view -h !{bam} | samtools view - -@ !{task.cpus} -F4 -q !{params.bam_mapping_quality_threshold} -o tmp_mapped.bam
+        samtools view -h !{bam} | samtools view - -@ !{task.cpus - 1} -f4 -o !{libraryid}.unmapped.bam
```
Comment (mahesh-panchal) [inline main.nf:1599]: Why is `samtools view` here piped to `samtools view` again?

**Review by mahesh-panchal — APPROVED** [review]

**Comment by mahesh-panchal** [comment]
> And for (2) I believe the `floor` part of the split_cpu definition is account for the divisible by 2 issue

Definitely a little embarrassed at not registering that.

**Review by apeltzer — APPROVED** [review]

File: main.nf (hunk 1 above)
Comment (apeltzer) [reply main.nf:1599]: Agreed, one could just run samtools view once with the `-h`flag and would get the same simplified 👍🏻

Comment (jfy133) [reply main.nf:1599]: Huh... no clue where that is coming from. will fix before merging!

File: main.nf [hunk 2]
```diff
@@ -1194,18 +1194,19 @@ process bwamem {
...
     """
-    bwa mem -t ${task.cpus} $fasta $r1 $r2 -R "@RG\\tID:ILLUMINA-${libraryid}\\tSM:${libraryid}\\tPL:illumina\\tPU:ILLUMINA-${libraryid}-${seqtype}" | samtools sort -@ ${task.cpus} -O bam - > "${libraryid}"_"${seqtype}".mapped.bam
+    bwa mem -t ${split_cpus} $fasta $r1 $r2 -R "@RG\\tID:ILLUMINA-${libraryid}\\tSM:${libraryid}\\tPL:illumina\\tPU:ILLUMINA-${libraryid}-${seqtype}" | samtools sort -@ ${split_cpus} -O bam - > "${libraryid}"_"${seqtype}".mapped.bam
     samtools index ${size} -@ ${task.cpus} "${libraryid}".mapped.bam
     """
     } else {
     """
-    bwa mem -t ${task.cpus} $fasta $r1 -R "@RG\\tID:ILLUMINA-${libraryid}\\tSM:${libraryid}\\tPL:illumina\\tPU:ILLUMINA-${libraryid}-${seqtype}" | samtools sort -@ ${task.cpus} -O bam - > "${libraryid}"_"${seqtype}".mapped.bam
-    samtools index -@ ${task.cpus} "${libraryid}"_"${seqtype}".mapped.bam ${size} 
+    bwa mem -t ${task.cpus} $fasta $r1 -R "@RG\\tID:ILLUMINA-${libraryid}\\tSM:${libraryid}\\tPL:illumina\\tPU:ILLUMINA-${libraryid}-${seqtype}" | samtools sort -@ ${split_cpus} -O bam - > "${libraryid}"_"${seqtype}".mapped.bam
```
Comment (mahesh-panchal) [inline main.nf:1208]: This is the one I think you wanted to change `bwa mem -t ${task.cpus}`.

File: main.nf [hunk 3]
```diff
@@ -1345,13 +1347,13 @@ process bowtie2 {
     //PE data without merging, PE data without any AR applied
     if ( seqtype == 'PE' && ( params.skip_collapse || params.skip_adapterremoval ) ){
     """
-    bowtie2 -x ${fasta} -1 ${r1} -2 ${r2} -p ${task.cpus} ${sensitivity} ${bt2n} ${bt2l} ${trim5} ${trim3} --maxins ${params.bt2_maxins} --rg-id ILLUMINA-${libraryid} --rg SM:${libraryid} --rg PL:illumina --rg PU:ILLUMINA-${libraryid}-${seqtype} 2> "${libraryid}"_bt2.log | samtools sort -@ ${task.cpus} -O bam > "${libraryid}"_"${seqtype}".mapped.bam
+    bowtie2 -x ${fasta} -1 ${r1} -2 ${r2} -p ${split_cpus} ${sensitivity} ${bt2n} ${bt2l} ${trim5} ${trim3} --maxins ${params.bt2_maxins} --rg-id ILLUMINA-${libraryid} --rg SM:${libraryid} --rg PL:illumina --rg PU:ILLUMINA-${libraryid}-${seqtype} 2> "${libraryid}"_bt2.log | samtools sort -@ ${split_cpus} -O bam > "${libraryid}"_"${seqtype}".mapped.bam
     samtools index "${libraryid}"_"${seqtype}".mapped.bam ${size}
```
Comment (mahesh-panchal) [inline main.nf:1351]: you use multiple cores above on L1203(old)/1204(new), but not here.

File: main.nf [hunk 4]
```diff
@@ -1345,13 +1347,13 @@ process bowtie2 {
...
     """
     } else {
     //PE collapsed, or SE data 
     """
-    bowtie2 -x ${fasta} -U ${r1} -p ${task.cpus} ${sensitivity} ${bt2n} ${bt2l} ${trim5} ${trim3} --rg-id ILLUMINA-${libraryid} --rg SM:${libraryid} --rg PL:illumina --rg PU:ILLUMINA-${libraryid}-${seqtype} 2> "${libraryid}"_bt2.log | samtools sort -@ ${task.cpus} -O bam > "${libraryid}"_"${seqtype}".mapped.bam
+    bowtie2 -x ${fasta} -U ${r1} -p ${split_cpus} ${sensitivity} ${bt2n} ${bt2l} ${trim5} ${trim3} --rg-id ILLUMINA-${libraryid} --rg SM:${libraryid} --rg PL:illumina --rg PU:ILLUMINA-${libraryid}-${seqtype} 2> "${libraryid}"_bt2.log | samtools sort -@ ${split_cpus} -O bam > "${libraryid}"_"${seqtype}".mapped.bam
     samtools index "${libraryid}"_"${seqtype}".mapped.bam ${size}
```
Comment (mahesh-panchal) [inline main.nf:1357]: similarly, multiple cores are not used here, but are above. There are more instances lower down as well.

File: main.nf [hunk 5]
```diff
@@ -2033,7 +2036,7 @@ process pmdtools {
     def platypus = params.pmdtools_platypus ? '--platypus' : ''
     """
     #Run Filtering step 
-    samtools calmd -b ${bam} ${fasta} | samtools view -h - | pmdtools --threshold ${params.pmdtools_threshold} ${treatment} ${snpcap} --header | samtools view -@ ${task.cpus} -Sb - > "${libraryid}".pmd.bam
+    samtools calmd -b ${bam} ${fasta} | samtools view -h - | pmdtools --threshold ${params.pmdtools_threshold} ${treatment} ${snpcap} --header | samtools view -@ ${task.cpus - 1} -Sb - > "${libraryid}".pmd.bam
```
Comment (mahesh-panchal) [inline main.nf:2039]: `samtools calmd -b ${bam} ${fasta} | samtools view -h - | pmdtools ...` also looks like another unnecessary use of `samtools view`. 

Also there are at least three tools in the pipe chain, but the `samtools view -@ ${task.cpus - 1}` at the end is using only one core less.

**Comment by jfy133** [comment]
@mahesh-panchal thanks for looking so closely. This was one of those cases where  you think 'oh this should be a quick PR', end up doing in a rush, and then miss lots of things :facepalm: .

I looked a bit more closely now and I _think_ I've caught everything.

**Review by mahesh-panchal — APPROVED** [review]

File: main.nf [hunk 6]
```diff
@@ -2035,15 +2035,15 @@ process pmdtools {
     def size = params.large_ref ? '-c' : ''
     def platypus = params.pmdtools_platypus ? '--platypus' : ''
     """
+    ## Note all `samtools view` steps ARE required for some reason
     #Run Filtering step 
-    samtools calmd -b ${bam} ${fasta} | pmdtools --threshold ${params.pmdtools_threshold} ${treatment} ${snpcap} --header | samtools view -@ ${task.cpus - 2} -Sb - > "${libraryid}".pmd.bam
+    samtools calmd -b ${bam} ${fasta} | samtools view -h - | pmdtools --threshold ${params.pmdtools_threshold} ${treatment} ${snpcap} --header | samtools view -Sb - > "${libraryid}".pmd.bam
```
[the remark this answers is not in the document]
Comment (jfy133) [reply main.nf]: Ok good call, I've tested and that seems to work. However, i've just checked the line history, and going to check with the actual writers of this process that it makes sense. I never used `pmdtools` mself :sweat_smile:

**Comment by jfy133** [comment]
@apeltzer @maxibor @TCLamnidis I've just changed the PMDtools command after suggestions from @mahesh-panchal . Can you check this looks OK with you? I never used `pmdtools` so I want to double check with you guys, who all actually have modified this line in the past according to git history.

**Comment by TCLamnidis** [comment]
I tried running a working command locally with and without samtools view, and it would throw an error when the `calmd` output was not passed through `samtools view`. 
```
$ samtools calmd -u JK2067_rmdup.bam hs37d5.fa | pmdtools --threshold 3 --header | samtools view -bo test.txt
Traceback (most recent call last):
  File "/Users/lamnidis/miniconda3/envs/pmdtools/bin/pmdtools", line 282, in <module>
    position = int(col[3])
IndexError: list index out of range
$ samtools calmd -u JK2067_rmdup.bam hs37d5.fa | samtools view -h - | pmdtools --threshold 3 --header | samtools view -bo test.txt
[bam_fillmd1] different MD for read 'K00233:65:HKTTVBBXX:6:2119:1387:28446': '30^T24' -> '30^N24'
[bam_fillmd1] different MD for read 'K00233:65:HKTTVBBXX:5:1112:4066:2070': '0A0C0T53' -> '0A0C0N53'
 $ samtools --version
samtools 1.3.1
Using htslib 1.3.1
Copyright (C) 2016 Genome Research Ltd.
```

**Comment by mahesh-panchal** [comment]
> Also what is -u? 
That tells samtools to output uncompressed bam. It's used when the next tool in the pipe can read bam, but saves time by not going through a compression-decompression step.

**Comment by jfy133** [comment]
```
❯  samtools --version
samtools 1.9
Using htslib 1.9
Copyright (C) 2018 Genome Research Ltd.
❯ samtools calmd JK2782_rmdup.bam Mammoth_MT_Krause.fasta | pmdtools --threshold 3 --CpG  --header | samtools view -Sb - > "JK2782".pmd.bam
❯ samtools calmd JK2782_rmdup.bam Mammoth_MT_Krause.fasta | pmdtools --deamination  --range 10 --CpG  -n 10000 > "JK2782".cpg.range."10".txt
```
This works for me? Is the `-u` really necessary?

**Comment by mahesh-panchal** [comment]
I don't think `pmdtools` can read bam, since it uses `samtools view` to produce SAM input.
https://github.com/pontussk/PMDtools

**Comment by jfy133** [comment]
Ok, then given default `calmd` output is `bam` anyway, (at least when I run it now), it's unecessary to specify.

Will proceed with merge, and someone can complain later if I brake something ;).

**Pull request merged**